### PR TITLE
feat: add support for QuoterV2

### PR DESCRIPTION
**Pull request by Jesse-Sawa**
The Uniswap/deploy-v3 (what we used for Celo) deploys the [V2 Quoter](https://github.com/Uniswap/deploy-v3/blob/main/src/steps/deploy-quoter-v2.ts#L1). Here is the [IQuoterV2 interface](https://github.com/Uniswap/swap-router-contracts/blob/main/contracts/interfaces/IQuoterV2.sol) to see the corresponding structs associated with these changes.

### CODE REVIEW

**Comment by JFrankfurt** [comment]
There's a lot of argument duplication here that makes the diff bigger than it needs to be. It looks like the only real differences in implementation are the argument wrapper and the string function names. Can you make the boolean logic handle only those decisions instead of wrapping duplicate argument constructions as well?

**Comment by JFrankfurt** [comment]
Can you also add some tests here that demonstrate usage of the new quoter option?

File: src/quoter.ts [hunk 1]
```diff
@@ -42,33 +48,37 @@ export abstract class SwapQuoter {
...
+        ...(options.useQuoterV2
+          ? tradeType === TradeType.EXACT_INPUT
+            ? { amountIn: quoteAmount }
+            : { amount: quoteAmount }
+          : { fee: route.pools[0].fee }),
+        ...(options.useQuoterV2 ? { fee: route.pools[0].fee } : { amount: quoteAmount }),
+        sqrtPriceLimitX96: toHex(options?.sqrtPriceLimitX96 ?? 0)
       }
+      const tradeTypeFunctionName =
+        tradeType === TradeType.EXACT_INPUT ? 'quoteExactInputSingle' : 'quoteExactOutputSingle'
```
Comment (JFrankfurt) [inline src/quoter.ts:67]: Would you mind doing the same thing for the call below?
(`quoteExactOutput` and `quoteExactInput`)

File: src/quoter.ts [hunk 2]
```diff
@@ -69,15 +68,15 @@ export abstract class SwapQuoter {
...
     if (singleHop) {
       const quoteV2Params: QuoteParams = {
-        amount: quoteAmount,
-        fee: route.pools[0].fee,
-        sqrtPriceLimitX96: toHex(options?.sqrtPriceLimitX96 ?? 0),
         tokenIn: route.tokenPath[0].address,
         tokenOut: route.tokenPath[1].address,
+        fee: route.pools[0].fee,
+        amount: quoteAmount,
+        sqrtPriceLimitX96: toHex(options?.sqrtPriceLimitX96 ?? 0)
```
Comment (JFrankfurt) [inline src/quoter.ts:75]: I didn't realize order matters here. It seems like that detail could easily be lost in future work. Can you add a comment, or better yet manually enforce order in your code?

File: src/quoter.ts [hunk 3]
```diff
@@ -42,34 +64,35 @@ export abstract class SwapQuoter {
     const singleHop = route.pools.length === 1
     const quoteAmount: string = toHex(amount.quotient)
     let calldata: string
+    const swapInterface = options.useQuoterV2 ? new Interface(IQuoterV2.abi) : this.INTERFACE
```
Comment (NoahZinsmeister) [inline src/quoter.ts:67]: this should be statically accessed and only constructed once. if you want to only do it when necessary, could write a custom get to cache the construction

File: src/quoter.ts (hunk 2 above)
Comment (NoahZinsmeister) [reply src/quoter.ts:75]: yeah the order matters and Object.values should not be relied upon, +1 to fixing this

File: src/quoter.ts [hunk 4]
```diff
@@ -42,34 +64,35 @@ export abstract class SwapQuoter {
...
-        ])
-      } else {
-        calldata = SwapQuoter.INTERFACE.encodeFunctionData(`quoteExactOutputSingle`, [
-          route.tokenPath[0].address,
-          route.tokenPath[1].address,
-          route.pools[0].fee,
-          quoteAmount,
-          toHex(options?.sqrtPriceLimitX96 ?? 0)
-        ])
+      const quoteV2Params: QuoteParams = {
```
Comment (NoahZinsmeister) [inline src/quoter.ts:70]: this is confusing, should be BaseQuoteParams and each v1/v2 subtype should not have optional amount members - can disambiguate the types down below and add the appropriate amount field depending on the quoter version

**Review by NoahZinsmeister — CHANGES_REQUESTED** [review]

**Review by JFrankfurt — APPROVED** [review]

**Review by NoahZinsmeister — APPROVED** [review]
lgtm

**Review by St0nersdash — COMMENTED** [review]
Viewed

**Pull request merged**